<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A quick start guide to smartid: Scoring and MARker selection method based on modified Tf-IDf • smartid</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A quick start guide to smartid: Scoring and MARker selection method based on modified Tf-IDf">
<meta property="og:description" content="smartid">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">smartid</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/smartid_Demo.html">A quick start guide to smartid: Scoring and MARker selection method based on modified Tf-IDf</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/DavisLaboratory/smartid/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A quick start guide to smartid: Scoring and
MARker selection method based on modified Tf-IDf</h1>
                        <h4 data-toc-skip class="author">Jinjin
Chen</h4>
            <address class="author_afil">
      Bioinformatics Division, Walter and Eliza Hall Institute of
Medical Research, Parkville, VIC 3052, AustraliaDepartment of Medical
Biology, University of Melbourne, Parkville, VIC 3010,
Australia<br><a class="author_email" href="mailto:#"></a><a href="mailto:chen.j@wehi.edu.au" class="email">chen.j@wehi.edu.au</a>
      </address>
                  
            <h4 data-toc-skip class="date">28 Mar 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DavisLaboratory/smartid/blob/HEAD/vignettes/smartid_Demo.Rmd" class="external-link"><code>vignettes/smartid_Demo.Rmd</code></a></small>
      <div class="hidden name"><code>smartid_Demo.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr>
<p><strong>smartid</strong></p>
<p><code>smartid</code> is a package that enables automated selection of
group specific signature genes, especially for rare population. This
package is developed for generating lists of specific signature genes
based on <strong>Term Frequency-Inverse Document Frequency</strong>
(TF-IDF) modified methods and <strong>expectation maximization</strong>
(EM) for labeled data. It can also be used as a new gene-set scoring
method or data transformation method for un-labeled data. Multiple
visualization functions are implemented in this package.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<hr>
<p><code>smartid</code> R package can be installed from Bioconductor or
<a href="https://github.com/DavisLaboratory/smartid" class="external-link">GitHub</a>.</p>
<p>The most updated version of <code>smartid</code> is hosted on GitHub
and can be installed using <code><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">devtools::install_github()</a></code>
function provided by <a href="https://cran.r-project.org/package=devtools" class="external-link">devtools</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"smartid"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"smartid"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prepare-data">Prepare Data<a class="anchor" aria-label="anchor" href="#prepare-data"></a>
</h2>
<hr>
<p>To show a quick start guide of <code>smartid</code>, here we use
package <em><a href="https://bioconductor.org/packages/3.19/splatter" class="external-link">splatter</a></em>
to simulate a scRNA-seq data of 1000 genes * 3000 cells. This data
consists of 4 groups, each has 2% DEGs except Group 4, which has no DEG
as a negative control group.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davislaboratory.github.io/smartid" class="external-link">smartid</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Oshlack/splatter" class="external-link">splatter</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sim_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/splatter/man/newParams.html" class="external-link">newSplatParams</a></span><span class="op">(</span></span>
<span>  nGenes <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  batchCells <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>  group.prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.4</span>, length.out <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,</span>
<span>  de.prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.02</span>, <span class="fl">0.02</span>, <span class="fl">0.02</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  <span class="co"># de.downProb = 0,</span></span>
<span>  de.facLoc <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  de.facScale <span class="op">=</span> <span class="fl">0.4</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/splatter/man/splatSimulate.html" class="external-link">splatSimulate</a></span><span class="op">(</span><span class="va">sim_params</span>, method <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get up markers based on fold change</span></span>
<span><span class="va">fc</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"DEFacGroup"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">data_sim</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">defac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">data_sim</span><span class="op">)</span><span class="op">[</span>, <span class="va">cols</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">up</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cols</span>, \<span class="op">(</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">defac</span>, <span class="fu">if_all</span><span class="op">(</span><span class="op">-</span><span class="op">!</span><span class="op">!</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span>, \<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">!</span><span class="op">!</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op">/</span> <span class="va">x</span> <span class="op">&gt;</span> <span class="va">fc</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">data_sim</span>, <span class="st">"metadata"</span><span class="op">)</span><span class="op">$</span><span class="va">up_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">up</span>, <span class="va">cols</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">data_sim</span>, <span class="st">"metadata"</span><span class="op">)</span><span class="op">$</span><span class="va">up_markers</span></span>
<span><span class="co">#&gt; $DEFacGroup1</span></span>
<span><span class="co">#&gt;  [1] "Gene31"  "Gene42"  "Gene172" "Gene225" "Gene308" "Gene312" "Gene352"</span></span>
<span><span class="co">#&gt;  [8] "Gene391" "Gene425" "Gene436" "Gene547" "Gene650" "Gene696" "Gene893"</span></span>
<span><span class="co">#&gt; [15] "Gene904" "Gene913"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DEFacGroup2</span></span>
<span><span class="co">#&gt;  [1] "Gene37"  "Gene76"  "Gene141" "Gene332" "Gene419" "Gene628" "Gene682"</span></span>
<span><span class="co">#&gt;  [8] "Gene713" "Gene778" "Gene818"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DEFacGroup3</span></span>
<span><span class="co">#&gt; [1] "Gene26"  "Gene28"  "Gene357" "Gene405" "Gene462" "Gene518" "Gene833"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $DEFacGroup4</span></span>
<span><span class="co">#&gt; character(0)</span></span>
<span><span class="va">data_sim</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 1000 3000 </span></span>
<span><span class="co">#&gt; metadata(2): Params up_markers</span></span>
<span><span class="co">#&gt; assays(6): BatchCellMeans BaseCellMeans ... TrueCounts counts</span></span>
<span><span class="co">#&gt; rownames(1000): Gene1 Gene2 ... Gene999 Gene1000</span></span>
<span><span class="co">#&gt; rowData names(8): Gene BaseGeneMean ... DEFacGroup3 DEFacGroup4</span></span>
<span><span class="co">#&gt; colnames(3000): Cell1 Cell2 ... Cell2999 Cell3000</span></span>
<span><span class="co">#&gt; colData names(4): Cell Batch Group ExpLibSize</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="labeled-data">Labeled Data<a class="anchor" aria-label="anchor" href="#labeled-data"></a>
</h2>
<hr>
<p><code>smartid</code> can be easily used to accurately identify
specific marker genes on labeled data. By adapting and modifying TF-IDF
approach, <code>smartid</code> shows robust power in finding marker
genes, especially for rare population which many methods fail in.</p>
<p><strong>marker identification of <code>smartid</code> includes 3 key
steps:</strong></p>
<p>step 1. score samples<br>
step 2. scale and transform scores<br>
step 3. identify markers using expectation maximization (EM)</p>
<div class="section level3">
<h3 id="score-samples">Score Samples<a class="anchor" aria-label="anchor" href="#score-samples"></a>
</h3>
<p>The first step is to score all samples/cells by using specified
approach. The score can be composed of 3 terms: TF (term/feature
frequency), IDF (inverse document/cell frequency) and IAE (inverse
average expression of features). Each term has a couple of available
choices with different formats to suit labeled or un-labeled data. Users
can use function <code><a href="../reference/idf_iae_methods.html">idf_iae_methods()</a></code> to see available methods
for IDF/IAE term. More details of each term can be seen in help page of
each function, e.g. <code>?smartid:::idf</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## show available methods</span></span>
<span><span class="fu"><a href="../reference/idf_iae_methods.html">idf_iae_methods</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;          unlabel HDBSCAN                label IGM              unlabel max </span></span>
<span><span class="co">#&gt;                    "hdb"                    "igm"                      "m" </span></span>
<span><span class="co">#&gt;                     null        label probability label relative frequency </span></span>
<span><span class="co">#&gt;                   "null"                   "prob"                     "rf" </span></span>
<span><span class="co">#&gt;               unlabel SD         unlabel standard </span></span>
<span><span class="co">#&gt;                     "sd"               "standard"</span></span></code></pre></div>
<p>The basic version of TF, IDF and IAE can be termed as:</p>
<p><span class="math inline">\(\mathbf{TF_{i,j}}=\frac{N_{i,j}}{\sum_j{N_{i,j}}},\)</span>
<span class="math inline">\(\mathbf{IDF_i} =
log(1+\frac{n}{n_i+1}),\)</span> <span class="math inline">\(\mathbf{IAE_i} = log(1+\frac{n}{\hat
N_{i,j}+1})\)</span></p>
<p>Where <span class="math inline">\(N_{i,j}\)</span> is the counts of
feature <span class="math inline">\(i\)</span> in cell <span class="math inline">\(j\)</span>; <span class="math inline">\(\hat
N_{i,j}\)</span> is <span class="math inline">\(max(0,N_{i,j}-threshold)\)</span>; <span class="math inline">\(n\)</span> is the total number of
documents(cells); <span class="math inline">\(n_i\)</span> is <span class="math inline">\(\sum_{j = 1}^{n} sign(N_{i,j} &gt;
threshold)\)</span>.</p>
<p>Here for labeled data, we can choose logTF * IDF_prob * IAE_prob for
marker identification.</p>
<p><span class="math display">\[\mathbf{score}=logTF*IDF_{prob}*IAE_{prob}\]</span></p>
<p>The probability version of IDF can be termed as: <span class="math inline">\(\mathbf{IDF_{i,j}} = log(1+\frac{\frac{n_{i,j\in
D}}{n_{j\in D}}}{max(\frac{n_{i,j\in \hat D}}{n_{j\in \hat D}})+
e^{-8}}\frac{n_{i,j\in D}}{n_{j\in D}})\)</span></p>
<p>And the probability version of IAE can be termed as: <span class="math inline">\(\mathbf{IAE_{i,j}} = log(1+\frac{mean(N_{i,j\in
D})}{max(mean(N_{i,j\in \hat D}))+ e^{-8}}*mean(N_{i,j\in D}))\)</span>
Where <span class="math inline">\(D\)</span> is the category of cell
<span class="math inline">\(j\)</span>; <span class="math inline">\(\hat
D\)</span> is the category other than <span class="math inline">\(D\)</span>.</p>
<p>TF here stands for gene frequency, which is similar to CPM, while IDF
represents the inverse cell/sample frequency for scRNA-seq data, and IAE
is the inverse average expression of each gene across all cells or cells
in each labeled group.</p>
<p>Another advantage of <code>smartid</code> is that it can start with
raw counts data, with no need for pre-processed data. And the scoring is
quite fast.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## compute score</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>  <span class="va">data_sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cal_score.html">cal_score</a></span><span class="op">(</span></span>
<span>    <span class="va">data_sim</span>,</span>
<span>    tf <span class="op">=</span> <span class="st">"logtf"</span>,</span>
<span>    idf <span class="op">=</span> <span class="st">"prob"</span>,</span>
<span>    iae <span class="op">=</span> <span class="st">"prob"</span>,</span>
<span>    par.idf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>label <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span>,</span>
<span>    par.iae <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>label <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   0.197   0.004   0.201</span></span>
<span></span>
<span><span class="co">## score and tf,idf,iae all saved</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">data_sim</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of length 7</span></span>
<span><span class="co">#&gt; names(7): BatchCellMeans BaseCellMeans BCV CellMeans TrueCounts counts score</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">data_sim</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Params"     "up_markers" "tf"         "idf"        "iae"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="scale-and-transform-score">Scale and Transform Score<a class="anchor" aria-label="anchor" href="#scale-and-transform-score"></a>
</h3>
<p>Scaling is needed to find the markers specific to the group, however,
standard scaling might fail due to the rare populations. Here
<code>smartid</code> uses a special scaling strategy
<code><a href="../reference/scale_mgm.html">scale_mgm()</a></code>, which can scale imbalanced data by given group
labels.</p>
<p>The score will be transformed using softmax before passing to EM
algorithm.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/top_markers.html">top_markers</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">data_sim</span>,</span>
<span>  label <span class="op">=</span> <span class="st">"Group"</span>,</span>
<span>  n <span class="op">=</span> <span class="cn">Inf</span> <span class="co"># set Inf to get all features processed score</span></span>
<span><span class="op">)</span></span>
<span><span class="va">top_m</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4,000 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   .dot [4]</span></span></span>
<span><span class="co">#&gt;    .dot   Genes    Scores</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Group1 Gene352 0.021<span style="text-decoration: underline;">6</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Group1 Gene696 0.020<span style="text-decoration: underline;">0</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Group1 Gene425 0.017<span style="text-decoration: underline;">1</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Group1 Gene391 0.016<span style="text-decoration: underline;">0</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Group1 Gene225 0.015<span style="text-decoration: underline;">9</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Group1 Gene893 0.014<span style="text-decoration: underline;">7</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Group1 Gene913 0.011<span style="text-decoration: underline;">3</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Group1 Gene650 0.010<span style="text-decoration: underline;">7</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Group1 Gene172 0.009<span style="text-decoration: underline;">98</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Group1 Gene547 0.009<span style="text-decoration: underline;">80</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3,990 more rows</span></span></span></code></pre></div>
<p>The top n features for each group will be ordered and listed in
<code>top_m</code>. <code>smartid</code> provides easy-to-use functions
to visualize top feature scores in each group and compare with actual
up-regulated DEGs.</p>
<p>It’s clear that the real UP DEGs are popping up to the top n
features. And for the negative control “Group 4”, the shape of top
feature score is totally different from the ones with DEGs, which can
provide more insights to help understand the data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/score_barplot.html">score_barplot</a></span><span class="op">(</span></span>
<span>  top_markers <span class="op">=</span> <span class="va">top_m</span>,</span>
<span>  column <span class="op">=</span> <span class="st">".dot"</span>,</span>
<span>  f_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">data_sim</span>, <span class="st">"metadata"</span><span class="op">)</span><span class="op">$</span><span class="va">up_markers</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<p>As we can see, there is an UP DEG ‘Gene76’ not popping up in Group 2,
we can check the relative expression of this gene using violin plot. It
is clear that this gene is not significantly highly expressed in Group 2
and the average expression is quite low across all cells.</p>
<p>This can also be confirmed in data simulation information, where the
scale factor is higher in Group2, but the GeneMean is too small to be
confident. Thus this gene won’t be selected by <code>smartid</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"Gene76"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">data_sim</span>, <span class="st">"metadata"</span><span class="op">)</span><span class="op">$</span><span class="va">tf</span><span class="op">[</span><span class="st">"Gene76"</span>, <span class="op">]</span>,</span>
<span>  Group <span class="op">=</span> <span class="va">data_sim</span><span class="op">$</span><span class="va">Group</span></span>
<span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html" class="external-link">geom_violin</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="va">Gene76</span>, fill <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## sim gene info</span></span>
<span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu">elementMetadata</span><span class="op">(</span><span class="va">data_sim</span><span class="op">)</span><span class="op">[</span><span class="fl">76</span>, <span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 1 row and 8 columns</span></span>
<span><span class="co">#&gt;          Gene BaseGeneMean OutlierFactor   GeneMean DEFacGroup1 DEFacGroup2</span></span>
<span><span class="co">#&gt;   &lt;character&gt;    &lt;numeric&gt;     &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1      Gene76   0.00628427             1 0.00628427           1     1.73743</span></span>
<span><span class="co">#&gt;   DEFacGroup3 DEFacGroup4</span></span>
<span><span class="co">#&gt;     &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1           1           1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="marker-selection">Marker Selection<a class="anchor" aria-label="anchor" href="#marker-selection"></a>
</h3>
<p>As we can see from above, there is a distinctly different
distribution of feature score between group with DEGs and without DEGs.
And there is a clear separation (break point) between the real DEGs and
non-DEGs.</p>
<p>To help automatically select real markers for each group,
<code>smartid</code> used an expectation maximization (EM) approach to
identify which genes fall into the real DEGs distribution and which are
not.</p>
<p>Regarding the distribution of scores as a mixture model, here we can
choose function <code><a href="../reference/markers_mixmdl.html">markers_mixmdl()</a></code> in <code>smartid</code> to
separate features. There are 2 available mixture model to choose: normal
(Gaussian) or gamma. We choose “norm” here as it runs faster.</p>
<p><code>smartid</code> also allows to plot the mixture distribution
plot after EM. It’s obvious that the top 2 components of Group 4 share
quite similar distribution, thus no markers will be selected for this
group.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">marker_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/markers_mixmdl.html">markers_mixmdl</a></span><span class="op">(</span></span>
<span>  top_markers <span class="op">=</span> <span class="va">top_m</span>,</span>
<span>  column <span class="op">=</span> <span class="st">".dot"</span>,</span>
<span>  ratio <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  dist <span class="op">=</span> <span class="st">"norm"</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; number of iterations= 88</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre></div>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<pre><code><span><span class="co">#&gt; number of iterations= 81</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-7-2.png" width="768"></p>
<pre><code><span><span class="co">#&gt; number of iterations= 36</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-7-3.png" width="768"></p>
<pre><code><span><span class="co">#&gt; number of iterations= 507</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-7-4.png" width="768"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marker_ls</span></span>
<span><span class="co">#&gt; $Group1</span></span>
<span><span class="co">#&gt;  [1] "Gene352" "Gene696" "Gene425" "Gene391" "Gene225" "Gene893" "Gene913"</span></span>
<span><span class="co">#&gt;  [8] "Gene650" "Gene172" "Gene547" "Gene436" "Gene904" "Gene312" "Gene308"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Group2</span></span>
<span><span class="co">#&gt; [1] "Gene419" "Gene332" "Gene141" "Gene37"  "Gene682" "Gene778" "Gene818"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Group3</span></span>
<span><span class="co">#&gt; [1] "Gene833" "Gene357" "Gene518" "Gene26"  "Gene405" "Gene462" "Gene28" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Group4</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>We can also compare our selected markers with real DEGs. As there is
no markers or DEG in group 4, only show overlap from Group1-3. It’s
clear that all the markers identified by <code>smartid</code> are real
DEGs, with a couple of missing genes in Group 1 and 2. But as what we
showed above, those genes only exhibit low mean expression across all
cells, thus not confident enough to be selected as markers.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/hms-dbmi/UpSetR" class="external-link">UpSetR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html" class="external-link">upset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html" class="external-link">fromList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="va">data_sim</span>, <span class="st">"metadata"</span><span class="op">)</span><span class="op">$</span><span class="va">up_markers</span>, <span class="va">marker_ls</span><span class="op">)</span><span class="op">)</span>, nsets <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-8-1.png" width="768"></p>
<p><code>smartid</code> also provides some other implementation of
marker selection. Here is another example using <code>mclust</code>.
Different from <code><a href="../reference/markers_mixmdl.html">markers_mixmdl()</a></code>,
<code><a href="../reference/markers_mclust.html">markers_mclust()</a></code> doesn’t need a pre-defined number of
components (which is 3 in <code><a href="../reference/markers_mixmdl.html">markers_mixmdl()</a></code>), instead, it
will select the number of components by searching a series of potential
numbers. This method is sometimes more robust than
<code><a href="../reference/markers_mixmdl.html">markers_mixmdl()</a></code>.</p>
<p>Similarly, this method also allows to plot the mixture distribution
for each component, but separately.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">marker_ls_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/markers_mclust.html">markers_mclust</a></span><span class="op">(</span></span>
<span>  top_markers <span class="op">=</span> <span class="va">top_m</span>,</span>
<span>  column <span class="op">=</span> <span class="st">".dot"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"max.one"</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-9-1.png" width="768"><img src="smartid_Demo_files/figure-html/unnamed-chunk-9-2.png" width="768"><img src="smartid_Demo_files/figure-html/unnamed-chunk-9-3.png" width="768"><img src="smartid_Demo_files/figure-html/unnamed-chunk-9-4.png" width="768"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">marker_ls_new</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">marker_ls_new</span><span class="op">)</span>, <span class="st">"new"</span><span class="op">)</span></span></code></pre></div>
<p>We can compare the marker list with the previous one. The overlap
result shows 2 methods can achieve almost the same marker list for each
group.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html" class="external-link">upset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html" class="external-link">fromList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">marker_ls</span>, <span class="va">marker_ls_new</span><span class="op">)</span><span class="op">)</span>, nsets <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="un-labeled-data">Un-labeled Data<a class="anchor" aria-label="anchor" href="#un-labeled-data"></a>
</h2>
<hr>
<p>While for the unlabeled data, <code>smartid</code> also provides the
score methods with no need for label information.</p>
<p>Here we choose logTF * IDF_sd * IAE_sd for for gene-set scoring as a
use case.</p>
<p><span class="math display">\[\mathbf{score}=logTF*IDF_{sd}*IAE_{sd}\]</span></p>
<p>Where <span class="math inline">\(\mathbf{IDF_i} =
log(1+SD(TF_{i})*\frac{n}{n_i+1})\)</span>, <span class="math inline">\(\mathbf{IAE_i} =
log(1+SD(TF_{i})*\frac{n}{\sum_{j=1}^{n}N_{i,j}+1})\)</span></p>
<div class="section level3">
<h3 id="score-samples-1">Score Samples<a class="anchor" aria-label="anchor" href="#score-samples-1"></a>
</h3>
<p>Similarly, the first step is to score samples/cells using the
specified method. This step also starts with raw counts data, without
need for data pre-processing, which is quite convenient and fast.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## compute score without label</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>  <span class="va">data_sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cal_score.html">cal_score</a></span><span class="op">(</span></span>
<span>    <span class="va">data_sim</span>,</span>
<span>    tf <span class="op">=</span> <span class="st">"logtf"</span>,</span>
<span>    idf <span class="op">=</span> <span class="st">"sd"</span>,</span>
<span>    iae <span class="op">=</span> <span class="st">"sd"</span>,</span>
<span>    new.slot <span class="op">=</span> <span class="st">"score_unlabel"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   0.181   0.016   0.197</span></span>
<span></span>
<span><span class="co">## new score is saved and tf,idf,iae all updated</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">data_sim</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of length 8</span></span>
<span><span class="co">#&gt; names(8): BatchCellMeans BaseCellMeans BCV ... counts score score_unlabel</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">data_sim</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Params"     "up_markers" "tf"         "idf"        "iae"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="compute-overall-score-for-gene-set">Compute Overall Score for Gene-set<a class="anchor" aria-label="anchor" href="#compute-overall-score-for-gene-set"></a>
</h3>
<p>To compare overall score of the given gene-set, we don’t need to
scale and transform score this time. Using <code><a href="../reference/gs_score.html">gs_score()</a></code> can
easily compute the overall score for each cell based on the given
gene-set list.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## compute score for each group marker list</span></span>
<span><span class="va">data_sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gs_score.html">gs_score</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">data_sim</span>,</span>
<span>  features <span class="op">=</span> <span class="va">marker_ls</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, <span class="co"># group 4 has no markers</span></span>
<span>  slot <span class="op">=</span> <span class="st">"score_unlabel"</span>,</span>
<span>  suffix <span class="op">=</span> <span class="st">"score.unlabel"</span> <span class="co"># specify the suffix of names to save</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## saved score</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sim</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Cell"                 "Batch"                "Group"               </span></span>
<span><span class="co">#&gt; [4] "ExpLibSize"           "Group1.score.unlabel" "Group2.score.unlabel"</span></span>
<span><span class="co">#&gt; [7] "Group3.score.unlabel"</span></span></code></pre></div>
<p>Now we get 3 columns of score for each group markers. We can then
visualize the score across groups, see how well it can discern the
target group.</p>
<p>It’s evident that the score can sufficiently separate the target
group from all others.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">data_sim</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="st">"Group1.score.unlabel"</span><span class="op">:</span><span class="st">"Group3.score.unlabel"</span>,</span>
<span>    names_to <span class="op">=</span> <span class="st">"group markers"</span>,</span>
<span>    values_to <span class="op">=</span> <span class="st">"score"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="va">score</span>, fill <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">`group markers`</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="smartid_Demo_files/figure-html/unnamed-chunk-13-1.png" width="960"></p>
</div>
</div>
<div class="section level2">
<h2 id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2024-03-24 r86185)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] UpSetR_1.4.0                scater_1.31.2              </span></span>
<span><span class="co">#&gt;  [3] scuttle_1.13.1              ggplot2_3.5.0              </span></span>
<span><span class="co">#&gt;  [5] splatter_1.27.0             SingleCellExperiment_1.25.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.33.3 Biobase_2.63.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.55.4        GenomeInfoDb_1.39.9        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.37.1              S4Vectors_0.41.5           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.49.1         MatrixGenerics_1.15.0      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.2.0           smartid_0.99.3             </span></span>
<span><span class="co">#&gt; [17] BiocStyle_2.31.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-7              gridExtra_2.3            </span></span>
<span><span class="co">#&gt;   [3] rlang_1.1.3               magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [5] tidytext_0.4.1            compiler_4.4.0           </span></span>
<span><span class="co">#&gt;   [7] DelayedMatrixStats_1.25.1 systemfonts_1.0.6        </span></span>
<span><span class="co">#&gt;   [9] vctrs_0.6.5               pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt;  [11] crayon_1.5.2              fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [13] backports_1.4.1           XVector_0.43.1           </span></span>
<span><span class="co">#&gt;  [15] labeling_0.4.3            utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [17] rmarkdown_2.26            ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt;  [19] ragg_1.3.0                purrr_1.0.2              </span></span>
<span><span class="co">#&gt;  [21] xfun_0.43                 zlibbioc_1.49.3          </span></span>
<span><span class="co">#&gt;  [23] cachem_1.0.8              beachmat_2.19.2          </span></span>
<span><span class="co">#&gt;  [25] jsonlite_1.8.8            SnowballC_0.7.1          </span></span>
<span><span class="co">#&gt;  [27] highr_0.10                DelayedArray_0.29.9      </span></span>
<span><span class="co">#&gt;  [29] BiocParallel_1.37.1       irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [31] parallel_4.4.0            R6_2.5.1                 </span></span>
<span><span class="co">#&gt;  [33] bslib_0.6.2               stringi_1.8.3            </span></span>
<span><span class="co">#&gt;  [35] jquerylib_0.1.4           Rcpp_1.0.12              </span></span>
<span><span class="co">#&gt;  [37] bookdown_0.38             knitr_1.45               </span></span>
<span><span class="co">#&gt;  [39] mixtools_2.0.0            splines_4.4.0            </span></span>
<span><span class="co">#&gt;  [41] Matrix_1.7-0              tidyselect_1.2.1         </span></span>
<span><span class="co">#&gt;  [43] abind_1.4-5               yaml_2.3.8               </span></span>
<span><span class="co">#&gt;  [45] viridis_0.6.5             codetools_0.2-19         </span></span>
<span><span class="co">#&gt;  [47] plyr_1.8.9                lattice_0.22-6           </span></span>
<span><span class="co">#&gt;  [49] tibble_3.2.1              withr_3.0.0              </span></span>
<span><span class="co">#&gt;  [51] evaluate_0.23             survival_3.5-8           </span></span>
<span><span class="co">#&gt;  [53] desc_1.4.3                mclust_6.1               </span></span>
<span><span class="co">#&gt;  [55] kernlab_0.9-32            pillar_1.9.0             </span></span>
<span><span class="co">#&gt;  [57] BiocManager_1.30.22       janeaustenr_1.0.0        </span></span>
<span><span class="co">#&gt;  [59] checkmate_2.3.1           plotly_4.10.4            </span></span>
<span><span class="co">#&gt;  [61] generics_0.1.3            RCurl_1.98-1.14          </span></span>
<span><span class="co">#&gt;  [63] sparseMatrixStats_1.15.0  munsell_0.5.0            </span></span>
<span><span class="co">#&gt;  [65] scales_1.3.0              glue_1.7.0               </span></span>
<span><span class="co">#&gt;  [67] lazyeval_0.2.2            tools_4.4.0              </span></span>
<span><span class="co">#&gt;  [69] data.table_1.15.2         BiocNeighbors_1.21.2     </span></span>
<span><span class="co">#&gt;  [71] tokenizers_0.3.0          ScaledMatrix_1.11.1      </span></span>
<span><span class="co">#&gt;  [73] locfit_1.5-9.9            fs_1.6.3                 </span></span>
<span><span class="co">#&gt;  [75] grid_4.4.0                tidyr_1.3.1              </span></span>
<span><span class="co">#&gt;  [77] colorspace_2.1-0          nlme_3.1-164             </span></span>
<span><span class="co">#&gt;  [79] GenomeInfoDbData_1.2.12   beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt;  [81] BiocSingular_1.19.0       vipor_0.4.7              </span></span>
<span><span class="co">#&gt;  [83] cli_3.6.2                 rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [85] textshaping_0.3.7         fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [87] segmented_2.0-3           S4Arrays_1.3.6           </span></span>
<span><span class="co">#&gt;  [89] viridisLite_0.4.2         dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [91] gtable_0.3.4              sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [93] digest_0.6.35             SparseArray_1.3.4        </span></span>
<span><span class="co">#&gt;  [95] ggrepel_0.9.5             htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt;  [97] farver_2.1.1              memoise_2.0.1            </span></span>
<span><span class="co">#&gt;  [99] htmltools_0.5.8           pkgdown_2.0.7            </span></span>
<span><span class="co">#&gt; [101] lifecycle_1.0.4           httr_1.4.7               </span></span>
<span><span class="co">#&gt; [103] MASS_7.3-60.2</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jinjin Chen.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
